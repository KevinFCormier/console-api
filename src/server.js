/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import https from 'https';
import fs from 'fs';
import schema from './schema/index';
import config from '../config';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const GRAPHQL_PORT = process.env.PORT || config.get('httpPort') || 4000;
const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(req => ({ formatError, schema, context: req })));

const privateKey = fs.readFileSync(process.env.serverKey || './sslcert/hcmuiapi.key', 'utf8');
const certificate = fs.readFileSync(process.env.serverCert || './sslcert/hcmuiapi.crt', 'utf8');
const credentials = { key: privateKey, cert: certificate };
const server = https.createServer(credentials, graphQLServer);

server.listen(GRAPHQL_PORT, () => {
  logger.info(`[pid ${process.pid}] [env ${process.env.NODE_ENV}] started.`);
  logger.info(`HCM UI API is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphql`);
  if (process.env.NODE_ENV !== 'production') {
    logger.info(`GraphiQL is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphiql`);
  }
});
