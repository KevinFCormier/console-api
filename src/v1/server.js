/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import schema from './schema/index';
import config from '../../config';
import HCMConnector from './datasource/lib/hcm-connector';
import MongoConnector from './datasource/lib/mongo-connector';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};


const getProductionOptions = async (req) => {
  const mongoConnector = new MongoConnector();
  const hcmConnector = new HCMConnector({
    hcmUrl: config.get('hcmUrl') || 'http://localhost:8080',
    pollInterval: config.get('hcmPollInterval') || 200,
    pollTimeout: config.get('hcmPollTimeout') || 10000,
  });

  await mongoConnector.connect(config.get('mongodbUrl') || 'mongodb://localhost:27017/weave');
  return {
    formatError,
    schema,
    context: {
      req,
      hcmConnector,
      mongoConnector,
    },
  };
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(getProductionOptions));

export default graphQLServer;
