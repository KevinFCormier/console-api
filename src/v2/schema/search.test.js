/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Search Resolver', () => {
  test('Correctly Resolves Search Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchResult: search(input: {keywords: [],
                filters: [ { property: "cluster", values: ["cluster1"]}]}){
              items
              related {
                kind
                count
                items
              }
              count
            }
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});
