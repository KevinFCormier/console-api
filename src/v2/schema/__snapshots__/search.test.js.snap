// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy Resolver Correctly Resolves Policy List Query 1`] = `
Object {
  "data": Object {
    "searchResult": Array [
      Object {
        "count": 100,
        "items": Object {
          "headers": Object {},
          "items": Array [
            Object {
              "cpu": "30%",
              "endpoint": "",
              "kind": "cluster",
              "memory": "20%",
              "name": "mock-cluster",
              "namespace": "mock-mcm-cluster",
              "nodes": 10,
              "status": "ok",
              "storage": "10%",
            },
            Object {
              "cpu": "30%",
              "endpoint": "",
              "kind": "cluster",
              "memory": "20%",
              "name": "mock-cluster-1",
              "namespace": "mock-mcm-cluster",
              "nodes": 10,
              "status": "ok",
              "storage": "10%",
            },
            Object {
              "cpus": 10,
              "kind": "node",
              "name": "mock-1.1.1.0",
              "role": "master",
            },
            Object {
              "cpus": 10,
              "kind": "node",
              "name": "mock-1.1.1.1",
              "role": "master",
            },
            Object {
              "cpus": 10,
              "kind": "node",
              "name": "mock-1.1.1.2",
              "role": "master",
            },
            Object {
              "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
              "hostIP": "179.160.35.59",
              "kind": "pod",
              "name": "mock-cluster",
              "namespace": "kube-system",
              "podIP": "179.160.35.59",
              "restarts": 20,
              "status": "Running",
            },
            Object {
              "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
              "hostIP": "179.160.35.59",
              "kind": "pod",
              "name": "mock-cluster-1",
              "namespace": "kube-system",
              "podIP": "179.160.35.59",
              "restarts": 20,
              "status": "Running",
            },
            Object {
              "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
              "hostIP": "179.160.35.59",
              "kind": "pod",
              "name": "mock-cluster-2",
              "namespace": "kube-system",
              "podIP": "179.160.35.59",
              "restarts": 20,
              "status": "Running",
            },
            Object {
              "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
              "hostIP": "179.160.35.59",
              "kind": "pod",
              "name": "mock-cluster-3",
              "namespace": "kube-system",
              "podIP": "179.160.35.59",
              "restarts": 20,
              "status": "Running",
            },
            Object {
              "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
              "hostIP": "179.160.35.59",
              "kind": "pod",
              "name": "mock-cluster-4",
              "namespace": "kube-system",
              "podIP": "179.160.35.59",
              "restarts": 20,
              "status": "Running",
            },
          ],
          "relatedResources": Array [
            Object {
              "count": 2,
              "kind": "application",
            },
            Object {
              "count": 3,
              "kind": "cluster",
            },
            Object {
              "count": 10,
              "kind": "deployment",
            },
            Object {
              "count": 200,
              "kind": "pod",
            },
          ],
        },
        "related": Array [
          Object {
            "count": 5,
            "kind": "cluster",
          },
          Object {
            "count": 3,
            "kind": "application",
          },
        ],
      },
    ],
  },
}
`;
